src/pip/_internal/resolution/resolvelib/base.py

```diff
@@ -143,8 +143,8 @@ def source_link(self):
         # type: () -> Optional[Link]
         raise NotImplementedError("Override in subclass")
 
-    def iter_dependencies(self, with_requires):
-        # type: (bool) -> Iterable[Optional[Requirement]]
+    def iter_dependencies(self, with_requires, requires_python):
+        # type: (bool, bool) -> Iterable[Optional[Requirement]]
```

Comment:
This is not needed. `Factory.make_requires_python_requirement()` already takes `Factory._ignore_requires_python` into account.

Reply:
All `ignore_requires_python` additions in Provider and Candidate classes should be removed.

Reply:
> `Factory.make_requires_python_requirement()` already takes `Factory._ignore_requires_python` into account.

This is true. However one of the goals of this PR is to avoid calling `Candidate._prepare()` when it's not necessary. See #1884 

Without my change, the following call chain happens:

```
Candidate.iter_dependencies()
-> Candidate._get_requires_python_dependency()
-> Candidate._prepare()   <<<<< heavy unnecessary work
-> Factory.make_requires_python_requirement()
-> if self._ignore_requires_python and return None
```

In other words, if this check was done earlier - we can avoid running candidate `setup.py` when `--ignore-requires-python` is given.

Another argument might be that this check doesn't belong to the `Factory`, and similar to `iter_dependencies(with_requires)` should be passed down to `Candidate` from `Provider.get_dependencies()`.

I'm going to change the PR to see what it looks like to move the flag out of `Factory` and into `Provider`.

Reply:
Avoiding source distribution build and consistency checks should be filed and discussed in a separate PR. Adding `--ignore-requires-python` to `download` is a straightforward change, and unrelated to this optimisation you are seeking here. If you want to add a flag to `download`, file a PR for it. If you want to avoid build time overhead in `download`, file a PR for it. Don’t do them in one.

Reply:
> Don’t do them in one.

Okay, I will submit a separate PR to add `--ignore-requires-python` to `pip download` and follow up with a separate PR  related to #1884 seeking to avoid preparing metadata when it's not necessary. Would this be ok?

For context: The ultimate motivation is to allow `pip download --no-deps --ignore-requires-python --no-binary :all: foo` to work without unpacking downloaded source distribution, because it's not needed. Please see https://github.com/pypa/pip/issues/1884#issuecomment-745568242 for more details.

Reply:
> Okay, I will submit a separate PR to add `--ignore-requires-python` to `pip download` and follow up with a separate PR […]. Would this be ok?

Yes, please do.